source/common/upstream/cluster_manager_impl.cc

```diff
@@ -487,9 +487,12 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
 
   if (existing_active_cluster != active_clusters_.end() ||
       existing_warming_cluster != warming_clusters_.end()) {
+    const auto cluster_it = (existing_active_cluster != active_clusters_.end())
```

Comment:
Not sure I follow where we add a cluster to init helper that isn't from active_clusters_.. do you have some code pointers?

Reply:
Clusters are loaded into the `warming_clusters_` map once all clusters are initialized. See https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L508-L510.

When a modification to a cluster is attempted prior to the cluster moving to the `active_clusters_` map, the undefined behavior access fixed above is encountered.

Reply:
Yep, but not spotting where this is added to `init_helper_` after this.

Reply:
Dug deeper based on your feedback and verified that warming clusters are not added to the `init_helper_`. I've modified the logic accordingly.